Tiered rejit work items

briansull opened the pull request:
1. Use info.compFullName as the input to create the JIT's MethodHash
2. Update JitOrder to print out MethodHash and PerfScore
3. Change eeGetMethodFullName to expand class and struct names for the argument types and the return type
4. fixed issue where bad edge weight were set in fgFoldConditional
- Made flEdgeWeightMin and flEdgeWeightMax private fields
- Added new method setNewEdgeWeights
- Added support method eeGetArgClass

BruceForstall commented on src/jit/block.h:
```diff
@@ -1274,13 +1274,23 @@ struct BasicBlockList
 
-    unsigned flDupCount; // The count of duplicate "edges" (use only for switch stmts)
+public:
+    BasicBlock::weight_t edgeWeightMin()
```
You can make these `const`

BruceForstall commented on src/jit/ee_il_dll.hpp:
```diff
@@ -109,6 +109,13 @@ inline var_types Compiler::eeGetArgType(CORINFO_ARG_LIST_HANDLE list, CORINFO_SI
+/*****************************************************************************/
+inline CORINFO_CLASS_HANDLE Compiler::eeGetArgClass(CORINFO_ARG_LIST_HANDLE list, CORINFO_SIG_INFO* sig)
+{
+    CORINFO_CLASS_HANDLE argClass = info.compCompHnd->getArgClass(sig, list);
```
Seems odd that you would reverse the arg order.

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -13082,6 +13080,14 @@ bool flowList::setEdgeWeightMaxChecked(BasicBlock::weight_t newWeight, BasicBloc
     return result;
 }
 
+void flowList::setNewEdgeWeights(BasicBlock::weight_t newMinWeight, BasicBlock::weight_t newMaxWeight)
```
Needs function header

BruceForstall commented on src/jit/block.h:
```diff
@@ -1289,13 +1299,14 @@ struct flowList
     //
     bool setEdgeWeightMinChecked(BasicBlock::weight_t newWeight, BasicBlock::weight_t slop, bool* wbUsedSlop);
     bool setEdgeWeightMaxChecked(BasicBlock::weight_t newWeight, BasicBlock::weight_t slop, bool* wbUsedSlop);
+    void setNewEdgeWeights(BasicBlock::weight_t newMinWeight, BasicBlock::weight_t newMaxWeight);
```
Could be just `setEdgeWeigths` -- no need for `New`.

briansull commented:
@BruceForstall  @dotnet/jit-contrib PTAL

BruceForstall commented:
I don't think you should add the JitOrderParser to the coreclr repo; I think you should put it in the jitutils repo instead. The jitOrderParser.cs file also needs a license header, and the filenames there should probably be all lower case to avoid future Linux issues.

jashook commented:
> I don't think you should add the JitOrderParser to the coreclr repo; I think you should put it in the jitutils repo instead. The jitOrderParser.cs file also needs a license header, and the filenames there should probably be all lower case to avoid future Linux issues.

I agree

jashook commented on src/tools/JitOrderParser/jitOrderParser.cs:
```diff
@@ -0,0 +1,333 @@
+
+        static void Main(string[] args)
+        {
+            int len = args.Length;
```
This seems to be missing usage documentation

briansull replied:
This tool is still just a work in progress. I will move it over to jitutils and work on it some more.

jashook commented on src/jit/compiler.cpp:
```diff
@@ -5584,31 +5589,42 @@ void Compiler::compCompileFinish()
-            printf(" mdToken |     |  RGN |    Count | EH | FRM | LOOP | NRM | IND | BBs | Cnt | Cnt | Cnt |  Alloc  |  IL  |   HOT |  COLD | method name \n");
-            printf("---------+-----+------+----------+----+-----+------+-----+-----+-----+-----+-----+-----+---------+------+-------+-------+-----------\n");
-            //      06001234 | PRF |  HOT |      219 | EH | ebp | LOOP |  15 |   6 |  12 |  17 |  12 |   8 |   28 p2 |  145 |   211 |   123 | System.Example(int)
+            printf("         |  Profiled   | Method   |   Method has    |   calls   | Num |LclV |AProp| CSE |   Perf  |bytes | %3s codesize| \n", Target::g_tgtCPUName);
```
What perfscore gets output on platforms that are missing full support like arm*?

briansull replied:
The default PerfScore algorithm counts each instruction executed as 1 unit.

jashook commented on src/tools/JitOrderParser/JitOrderParser.csproj:
```diff
@@ -0,0 +1,8 @@
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
I think at this point we should target netcoreapp3.0

briansull replied:
OK

jashook commented on src/tools/JitOrderParser/JitOrderParser.sln:
```diff
@@ -0,0 +1,25 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
I think the general suggestion is to not check in sln files.

briansull replied:
OK

briansull commented:
@BruceForstall @jashook @dotnet/jit-contrib PTAL
I have addressed the feedback from Bruce and Jarret

With this change we will get unique hashes for all methods that have different signatures.

AndyAyersMS commented:
Seems like the edge weight changes are unrelated to the rest and should be a separate PR. I'd hoped that we could get rid of the min/max weight scheme as it's just confusing. Why bother having two weights? Which one is more trustworthy?

I wonder if the hash changes resolve #1957? Do we now get unique hashes for IL stubs?

There are a number of places already where we call `getArgClass` directly. I never really understood the point of the wrappers for jit interface calls but it seems like we should be consistent and either always directly call or always call via a wrapper.

briansull commented:
> I wonder if the hash changes resolve #1957? Do we now get unique hashes for IL stubs?

Yes, IL Stubs are uniquely identified with the new methodhash

briansull commented:
I ran into an assert with edge weights when doing IBC training and optimization using a checked build.
This edge weight work was done to allow me to find and fix the bug.
I would like to get this work checked in, redesign is a separate issue here.

briansull commented:
@AndyAyersMS if you really want I can checked these items in separately.

briansull commented on src/jit/morph.cpp:
```diff
@@ -14975,22 +14974,34 @@ bool Compiler::fgFoldConditional(BasicBlock* block)
+                            edge         = fgGetPredForBlock(bUpdated->bbNext, bUpdated);
+                            newMaxWeight = bUpdated->bbWeight;
+                            newMinWeight = min(edge->edgeWeightMin(), newMaxWeight);
+                            edge->setEdgeWeights(newMinWeight, newMaxWeight);
```
These assignments of flEdgeWeightMax  could violate the requirements that  min-weight <= maxWeight

AndyAyersMS commented:
> @AndyAyersMS if you really want I can checked these items in separately

I think it's a good practice not to mix unrelated (or seemingly unrelated) changes in one PR, but I'm not going to ask you to split this one up, as the rest of the changes are diagnostic.

We don't have a good mechanism for creating repros or regression tests for IBC related bugs. Seems like we ought to figure out how to do this.

briansull commented:
> We don't have a good mechanism for creating repros or regression tests for IBC related bugs. Seems like we ought to figure out how to do this.

I agree that we should have a better way to create regression repros in this area.  
(Checking in the IBC data and running the IbcMerge tool are required steps)
I believe that their is some rework going on in the profile area,
when that comes online we should have the ability to added regression testing.

briansull commented:
@BruceForstall  @jashook  @dotnet/jit-contrib PTAL

 This is ready for checkin

BruceForstall commented:
> There are a number of places already where we call getArgClass directly. I never really understood the point of the wrappers for jit interface calls but it seems like we should be consistent and either always directly call or always call via a wrapper.

There's no need for a wrapper unless the wrapper does additional work. In many cases, we use wrappers so the altjit case can do something different.
